packages/flutter/lib/src/material/theme_data.dart

```diff
@@ -824,6 +848,117 @@ class ThemeData extends Diagnosticable {
+///
+/// A [Map] from Material widget types to a boolean describes whether the
+/// adaptation should be made for widgets that is wrapped with [PlatformBuilder]
+/// descendant to this [Theme] and whose [PlatformBuilder.themeAdaptiveType]
+/// matches the type in the [adaptiveWidgets] [Map].
+///
+/// When a widget type is not explicitly specified in [adaptiveWidgets], the
+/// [defaultAdaptiveness] value is used.
+///
+/// The singleton [none] instance can be used to signal no adaptation on any
+/// descendent widgets.
+///
+/// The singleton [bundled] instance can be used to signal adaptation on all
+/// Material widgets.
+///
+/// The singleton [all] instance can be used to signal adaptation on all widgets
+/// including user-custom usages of [PlatformBuilder].
+///
+/// Individual [PlatformBuilder] descendents can still override this [Theme]
+/// by specifying null for [PlatformBuilder.themeAdaptiveType].
+@immutable
+class AdaptiveWidgetThemeData extends Diagnosticable{
+  /// Create a [AdaptiveWidgetThemeData] for a [Theme] that instructs whether
+  /// descendent widgets should adapt to the platform.
+  ///
+  /// The parameters [adaptiveWidgets] and [defaultAdaptiveness] cannot be null.
+  const AdaptiveWidgetThemeData(
+    this.adaptiveWidgets, {
+    this.defaultAdaptiveness,
+  }) : assert(adaptiveWidgets != null);
+
+  /// A [AdaptiveWidgetThemeData] where none of the descendents dynamically
+  /// adapt to the platform unless overridden in [PlatformBuilder].
+  static const AdaptiveWidgetThemeData none =
+      const AdaptiveWidgetThemeData(const <Type, bool>{});
+
+  /// A [AdaptiveWidgetThemeData] where all bundled Material Design widgets
+  /// that support platform adaptations will adapt when running on iOS.
+  static const AdaptiveWidgetThemeData bundled =
+      const AdaptiveWidgetThemeData(const <Type, bool>{
+        FlatButton: true,
+        RaisedButton: true,
+        Slider: true,
+        Switch: true,
+      });
+
+  /// A [AdaptiveWidgetThemeData] where all of the descendents dynamically
+  /// adapt to the platform if supported unless overridden in [PlatformBuilder].
+  static const AdaptiveWidgetThemeData all =
+      const AdaptiveWidgetThemeData(const <Type, bool>{}, defaultAdaptiveness: true);
+
+  /// A [Map] from widget type to whether descendents of that widget type should
+  /// dynamically adapt to the platform.
+  ///
+  /// If the widget type is not in the map, or if the corresponding value is
+  /// null, the value of [defaultAdaptiveness] will be used.
+  ///
+  /// The map itself cannot be null.
+  final Map<Type, bool> adaptiveWidgets;
+
+  /// A boolean indicating the default adaptiveness of a widget type if it's
+  /// not explicitly specified in [adaptiveWidgets].
+  ///
+  /// It can be used to turn on or off the adaptiveness of all widget types
+  /// without specifying them individually in [adaptiveWidgets].
+  ///
+  /// The value defaults to [false].
```

Comment:
the constructor defaults it to null, not false

Reply:
It ends up being interpreted as false. I noticed this doc pattern in a few places where we describe a deeper defaulting value when constructed with null. Let me know if you prefer those to be described differently.

Reply:
unless there's a strong reason to do otherwise, we should put the defaulting in the constructor rather than on use. it's slightly more efficient, more self-documenting, will be more compatible with non-nullable types in due course, and is less likely to be buggy due to different places defaulting to different values.

Reply:
ya, I did it initially though I removed the null check to be able to do subtree partial overrides without explicitly reading back out the values that are inherited and re-inserting them back in the constructor. I could get around it by changing the `merge` signature to be something like `mergeWithCopyWithFromTheseFields` but it seems to depart from patterns used in other places :S

Reply:
Not sure I understand what you mean by partial overrides. Can you elaborate?

Reply:
Might be easier to explain in https://github.com/flutter/flutter/pull/17823/files#diff-e0e4aef66f23cfc137ba8d0976a2d95fR178. Assume those Themes are in completely different files, this makes it easier to only override one thing without affecting anything else. 

Though maybe there's a better way of doing it.